db/view/row_locking.cc

```diff
@@ -73,32 +73,31 @@ future<row_locker::lock_holder>
 row_locker::lock_pk(const dht::decorated_key& pk, bool exclusive, db::timeout_clock::time_point timeout, stats& stats) {
     mylog.debug("taking {} lock on entire partition {}", (exclusive ? "exclusive" : "shared"), pk);
     auto tracker = latency_stats_tracker(exclusive ? stats.exclusive_partition : stats.shared_partition);
+    auto holder = lock_holder(this);
     auto i = _two_level_locks.try_emplace(pk, this).first;
-    auto f = exclusive ? i->second._partition_lock.write_lock(timeout) : i->second._partition_lock.read_lock(timeout);
-    co_await std::move(f);
+    co_await holder.lock_partition(&i->first, i->second._partition_lock, exclusive);
     // Note: we rely on the fact that &i->first, the pointer to a key, never
     // becomes invalid (as long as the item is actually in the hash table),
     // even in the case of rehashing.
     tracker.lock_acquired();
-    co_return lock_holder(this, &i->first, exclusive);
+    co_return holder;
 }
 
 future<row_locker::lock_holder>
 row_locker::lock_ck(const dht::decorated_key& pk, const clustering_key_prefix& cpk, bool exclusive, db::timeout_clock::time_point timeout, stats& stats) {
     mylog.debug("taking shared lock on partition {}, and {} lock on row {} in it", pk, (exclusive ? "exclusive" : "shared"), cpk);
     auto tracker = latency_stats_tracker(exclusive ? stats.exclusive_row : stats.shared_row);
+    auto holder = lock_holder(this);
     auto i = _two_level_locks.try_emplace(pk, this).first;
-    auto partition_lock_holder = co_await i->second._partition_lock.hold_read_lock(timeout);
+    co_await holder.lock_partition(&i->first, i->second._partition_lock, false);
     auto j = i->second._row_locks.find(cpk);
     if (j == i->second._row_locks.end()) {
         // Not yet locked, need to create the lock. This makes a copy of cpk.
         j = i->second._row_locks.emplace(cpk, lock_type()).first;
     }
-    auto row_lock_holder = co_await (exclusive ? j->second.hold_write_lock(timeout) : j->second.hold_read_lock(timeout));
+    co_await holder.lock_row(&j->first, j->second, exclusive);
```

Comment:
I see we currently do 

```c++
    return when_all_succeed(std::move(lock_partition), std::move(lock_row))
    .then_unpack([this, pk = &i->first, cpk = &j->first, exclusive, &single_lock_stats, waiting_latency = std::move(waiting_latency)] (auto lock1, auto lock2) mutable {
        lock1.release();
        lock2.release();
```

which looks bad. But isn't it sufficient to move lock1 and lock2 into a lock_holder object? Seems simpler overall.

Reply:
It can be done, but will require some changes to class lock_holder so it can be constructed in advance and then accept the individual lock holders once they are acquired, or as I did in this series, using lock_holder methods to acquire the locks and then store whatever is required in the lock_holder object to unlock once destroyed.